google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -3321,6 +3399,26 @@ private boolean canCreateSession() {
     }
   }
 
+  private void maybeCreateMultiplexedSession() {
+    synchronized (lock) {
+      if (!multiplexedSessionBeingCreated) {
+        logger.log(Level.FINE, String.format("Creating multiplexed sessions"));
+        try {
+          multiplexedSessionBeingCreated = true;
+          sessionClient.createMultiplexedSession(multiplexedSessionConsumer);
+        } catch (Throwable ignore) {
+          // such an exception will never be thrown. the exception will be passed onto the consumer.
+        }
+      }
+    }
+  }
+
+  private void handleCreateMultiplexedSessionFailure(SpannerException e) {
+    synchronized (lock) {
+      multiplexedSessionFuture.get().setException(e);
```

Comment:
I'm wondering what the best strategy is here:
1. For the initial session creation: This is the way we should do it.
2. For session refreshing, we have to options:
   2.1. Set the exception here. That means that all operations from this point on will fail, as there is no longer a valid multiplexed session, and this error will be propagated to the application.
   2.2. Only log the error, and let the maintainer retry creating a new session. The old multiplexed session will then continue to be used by the application, and this will work as long as it does not expire completely.

I lean slightly towards option 2.2 for refreshing the session, as it does not break the application right away if session refresh fails once.

Reply:
Yes, I have now separated two consumers `MultiplexedSessionInitializationConsumer` and `MultiplexedSessionMaintainerConsumer` and achieved this behavior. Also as discussed added a 10 minutes delay in the maintainer task to ensure we don't retry too often in case of such failures when a session is already available.